### PR TITLE
[tmva][sofie] Enhance Softplus stability in ROperator_BasicUnary

### LINKED ISSUE #21071: [tmva][sofie] Improve numerical stability and overflow protection for Softplus operator
### Explain what you would like to see improved and how.

The current Softplus implementation (recently merged via #21050) relies on the `ROperator_BasicUnary` trait with the standard formula `std::log(std::exp(x) + 1)`. While functionally correct for small inputs, this implementation has two significant numerical limitations in edge cases for scientific computing:

* **Overflow:** For inputs $x > 88.72$ (in float32), `std::exp(x)` overflows to infinity, causing the operator to return `inf`. The expected behavior for large inputs is linear ($f(x) \approx x$).
* **Precision Loss:** For large negative inputs, `exp(x)` becomes very small. Computing `1 + exp(x)` discards significant digits before the log calculation, leading to precision errors.

Leading inference frameworks (PyTorch, TensorFlow) mitigate this by using a threshold for the linear region and `log1p` for the curve.

## Proposal
I propose updating the Softplus logic to use numerical stability safeguards:

1.  **Thresholding:** For $x \ge 20.0f$, return $x$.
    * At $x=20$, $\ln(1+e^{20}) \approx 20.000000002$, which is indistinguishable from $20.0$ in float32.
    * This prevents the `exp()` overflow completely for large positive values.
2.  **Precision:** For $x < 20.0f$, use `std::log1p(std::exp(x))`.
    * `std::log1p(z)` computes $\ln(1+z)$ accurately even when $z \approx 0$, preserving precision for negative inputs.

### Proposed Logic:
```cpp
// Threshold 20.0f defined as exact hexfloat 0x1.4p+4f
return (x >= 20.0f) ? x : std::log1p(std::exp(x));
```
## Action Plan
I have a complete implementation ready (from PR #21023) that includes:

* **Bit-exact Hexfloat constants** for the threshold.
* **Comprehensive GTest suite** (`TestSofieSoftplus.cxx`) verifying:
    * **Overflow protection** (inputs > 89 return $x$, not `Inf`).
    * **Stability** in the linear region.
    * **Correct usage** of `std::log1p` vs `std::log`.

I can submit a Pull Request to patch the `ROperator_BasicUnary` implementation to include this logic and integrate the validation suite.

### ROOT version

master 

### Installation method

build from source

### Operating system

Linux (Ubuntu 24.04)

### Additional context

**References & Prior Art:**
* **PyTorch Standard:** PyTorch's `Softplus` implementation utilizes a `threshold` parameter (default 20) to revert to a linear function for large inputs to avoid overflow.
  * *Source:* https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html
* **TensorFlow:** Similarly uses `softplus(x) = x` for `x > threshold`.

**Related PRs:**
* **Current Implementation:** #21050 (Merged) - Established the operator.
* **Proposed Code Source:** #21023 (Closed) - I have a fully tested branch with the `log1p` and `threshold` logic ready to port.

cc: @lmoneta @bellenot @sanjibansg

### PR DESCRIPTION
**Pull request by AdityaDRathore**
### Description
This PR implements numerical stability improvements for the `Softplus` operator, addressing overflow issues identified in Issue #21071.

It supersedes PR #21023 (which was closed due to branch history alignment).

### Changes
* **Hexfloat Thresholding:** Implements a hard threshold check using `0x1.4p+4f` (20.0f) to prevent `exp()` overflow for large inputs.
* **Precision Logic:** Replaces the naive `log(exp(x)+1)` with `std::log1p(std::exp(x))` to preserve precision for negative inputs.
* **Validation:** Adds a dedicated unit test suite `TestSofieSoftplus.cxx` verifying bit-exact hexfloat emission and overflow protection.

### Technical Context
* **Target:** `ROperator_BasicUnary` (via `UnaryOpTraits` specialization).
* **Context:** This implementation follows the approach suggested by @sanjibansg in Issue #21071 to patch the existing architecture rather than introducing a new operator class.

Fixes #21071

cc: @sanjibansg @lmoneta @bellenot

### CODE REVIEW

**Review by sanjibansg — CHANGES_REQUESTED** [review]
Hi @AdityaDRathore,
Thanks for this PR. Can you add the tests following the convention we have in the SOFIE sub-directory?
https://github.com/root-project/root/tree/master/tmva/sofie/test

**Comment by AdityaDRathore** [comment]
Hi @sanjibansg, thanks for the review

I've added tests following the convention in the Sofie test directory, as you mentioned.

File: tmva/sofie/test/TestSofieSoftplus.cxx [hunk 1]
```diff
@@ -0,0 +1,191 @@
+#include "TMVA/ROperator_BasicUnary.hxx"
```
Comment (sanjibansg) [inline tmva/sofie/test/TestSofieSoftplus.cxx]: Can we follow the unit test convention like we have in for the tests you are proposing here: https://github.com/AdityaDRathore/root/blob/2021a29910999c85761fb2caef9339220bcbbc72/tmva/sofie/test/TestCustomModelsFromONNX.cxx

**Review by sanjibansg — DISMISSED** [review]
Hi @AdityaDRathore,
Thanks for this implementation. I have one comment.

**Comment by AdityaDRathore** [comment]
Hi @sanjibansg, thanks for it. I've cleaned up the tests and files.

**Review by lmoneta — COMMENTED** [review]
LGTM

**Review by guitargeek — APPROVED** [review]
Thanks! I rebased on top of `master`, and will merge once the tests are green.

**Comment by AdityaDRathore** [comment]
Thanks Jonas!
